# Week 8 Independent Lab: Advanced Data Manipulation

Author: Hali Evans    
Course: BGEN632    
Purpose: Practice advanced functions in pandas to query and filter data.

In [6]:
# import modules
import os
import pandas as pd

In [7]:
# set working directory
os.chdir("/Users/halievans/Documents/BGEN632/week8labs/data")  # add your filepath within the quotes
os.getcwd()  # confirm change

'/Users/halievans/Documents/BGEN632/week8labs/data'

In [8]:
# load data
gh_tfp_df = pd.read_csv("gh_turnover_four_projects.csv")
gh_tfp_df.info()  # quick inspect

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 59.3+ KB


## Rename Columns

In [10]:
# rename columns
gh_rename_df = (
    gh_tfp_df
    .rename(columns={
        "domain" : "project_domain",
        "language" : "project_language",
        "windows" : "project_age",
        "window_idx" : "quarter",
        "num_team" : "team_size",
        "num_commits" : "project_commits",
        "blau_gender" : "gender_blau",
        "Gini_gh_ten" : "gh_tenure_gini", 
        "Core1" : "core_dev",
        "commits" : "user_commits",
        "propCommits" : "commit_proportion",
        "github_tenure" : "gh_tenure",
        "leavesNextQ" : "leaves_next_q"})
)
gh_rename_df.head()  # display first n rows

project_id project_domain project_language  project_age  quarter  \
0        1101          OTHER           Python           23       22   
1        1101          OTHER           Python           23       22   
2        1101          OTHER           Python           23       22   
3        1101          OTHER           Python           23       22   
4        1101          OTHER           Python           23       22   

   team_size  project_commits  gender_blau  gh_tenure_gini  user_id  core_dev  \
0         26            22688          0.0        0.422259  1614129         1   
1         26            22688          0.0        0.422259  1614101         0   
2         26            22688          0.0        0.422259    94412         0   
3         26            22688          0.0        0.422259    88459         0   
4         26            22688          0.0        0.422259   738787         0   

   user_commits  commit_proportion gender  gh_tenure  leaves_next_q  
0         22138           0.975758   male        426              0  
1           124           0.005465   male         46              0  
2            53           0.002336   male        941              1  
3            49           0.002160   male       1870              0  
4            45           0.001983   male        781              0

## Data Wrangling

In [12]:
# calculate the mean of user_commits
mean_uc = gh_rename_df["user_commits"].mean() # calculation
print("Mean:", mean_uc) # display mean

Mean: 339.36575052854124


In [13]:
# calculate the median of user_commits
median_uc = gh_rename_df["user_commits"].median() # calculation
print("Median:", median_uc) # display median

Median: 18.0


In [14]:
# select columns that start with a p or contain a g in new dataframe and output columns
turnover_new = (
    gh_rename_df
    .filter(regex="(^p)|g") # select desired columns
)
turnover_new.columns

Index(['project_id', 'project_domain', 'project_language', 'project_age',
       'project_commits', 'gender_blau', 'gh_tenure_gini', 'gender',
       'gh_tenure'],
      dtype='object')

In [15]:
# select project_commits greater than 9000
(turnover_new
 .query("project_commits > 9000")  # keep rows based on condition
 .head() # display first n rows
)

project_id project_domain project_language  project_age  project_commits  \
0        1101          OTHER           Python           23            22688   
1        1101          OTHER           Python           23            22688   
2        1101          OTHER           Python           23            22688   
3        1101          OTHER           Python           23            22688   
4        1101          OTHER           Python           23            22688   

   gender_blau  gh_tenure_gini gender  gh_tenure  
0          0.0        0.422259   male        426  
1          0.0        0.422259   male         46  
2          0.0        0.422259   male        941  
3          0.0        0.422259   male       1870  
4          0.0        0.422259   male        781

In [16]:
# using the filter and query above use pandas piping notation to perform the operations together in new dataframe.
turnover_newer = (
    gh_rename_df
    .filter(regex="(^p)|g")  # select desired columns
    .query("project_commits > 9000")  # keep rows based on condition
)
turnover_newer.head() # display first n rows

project_id project_domain project_language  project_age  project_commits  \
0        1101          OTHER           Python           23            22688   
1        1101          OTHER           Python           23            22688   
2        1101          OTHER           Python           23            22688   
3        1101          OTHER           Python           23            22688   
4        1101          OTHER           Python           23            22688   

   gender_blau  gh_tenure_gini gender  gh_tenure  
0          0.0        0.422259   male        426  
1          0.0        0.422259   male         46  
2          0.0        0.422259   male        941  
3          0.0        0.422259   male       1870  
4          0.0        0.422259   male        781

In [17]:
# using pandas piping select columns that end with letter e or contain letter q select rows in which gh_tenure is greater than or equal to 1991
gh_filter_df = (
    gh_rename_df
    .filter(regex="e$|q")  # select desired columns
    .query("gh_tenure >= 1991")  # keep rows based on condition
)
gh_filter_df.head() # display first n rows

project_language  project_age  quarter  team_size  gh_tenure  \
339             Java           16       17          2       2006   
457           Python           23       24         20       2054   

     leaves_next_q  
339              0  
457              0

In [18]:
# add user commits to filter
gh_filter_df = (
    gh_rename_df
    .filter(regex="e$|q|^user_commits$")  # select desired columns
    .query("gh_tenure >= 1991")  # keep rows based on condition
)
gh_filter_df.head()

project_language  project_age  quarter  team_size  user_commits  \
339             Java           16       17          2             6   
457           Python           23       24         20             1   

     gh_tenure  leaves_next_q  
339       2006              0  
457       2054              0

In [19]:
mean_filter_uc = gh_filter_df["user_commits"].mean() # calculation
median_filter_uc = gh_filter_df["user_commits"].median() # calculation

print("Mean:", mean_filter_uc) # display mean
print("Median:", median_filter_uc) # display median

Mean: 3.5
Median: 3.5


In [20]:
gh_rename_df.shape[0]

473

#### Original Calculations
Mean: 339.36575052854124    
Median: 18.0    

#### Filtered Calculations
Mean: 3.5    
Median: 3.5    

It is important to note that the data was reduced from 473 rows to just 2. As a result, the mean and median values in the filtered dataset are significantly lower than those in the original. This difference stems from what was chosen to filter the dataset and is a reminder to analyze data prior to applying a filter to ensure there are not any unintended consequences.